### PR TITLE
Fix root group subgroup bug

### PR DESCRIPTION
**Pull request by filipelbc**
Closes: #30
Was not listing group/subgroup "" or "id 0"

Python strips arguments, so that `""` (empty string) is not passable as argument. Use `"."` instead.

En passant:
Closes: #34

Also some small refactor, some tests and some makefile improvement.

### CODE REVIEW

File: dit/dit.py [hunk 1]
```diff
@@ -240,25 +269,28 @@ def _load_index(self):
...
                 if len(p) == 0:
-                    group, subgroup = None, None
+                    group, subgroup = self.root_name, self.root_name
                 elif len(p) == 1:
-                    group, subgroup = None, p[0]
-                else:
+                    group, subgroup = self.root_name, p[0]
+                elif len(p) == 2:
                     group, subgroup = p
+                else:
```
Comment (danielmoraes) [inline dit/dit.py:288]: As far as I know, len(p) will never be greater than 2. Why the change?

File: dit/dit.py [hunk 2]
```diff
@@ -327,90 +359,73 @@ def _export_all(self, concluded, verbose):
...
-        ids = list(map(int, [i if i else "0" for i in ids]))
         if len(ids) > 3:
-            raise Exception("Invalid GID format")
+            raise Exception("Invalid ID format")
+        ids = [int(i) if i else "0" for i in ids]
         ids = [None] * (3 - len(ids)) + ids
         group_id, subgroup_id, task_id = ids
 
         if group_id is None:
-            group_id = 0
```
Comment (danielmoraes) [inline dit/dit.py:406]: This line is necessary when there is not a current group set (unless you change current_group to "").

File: dit/dit.py [hunk 3]
```diff
@@ -327,90 +359,73 @@ def _export_all(self, concluded, verbose):
...
             for i in range(len(self.index)):
                 if self.index[i][0] == self.current_group:
                     group_id = i
                     break
 
         if subgroup_id is None:
-            subgroup_id = 0
```
Comment (danielmoraes) [inline dit/dit.py:413]: This line is necessary when there is not a current subgroup set  (unless you change current_subgroup to "").

File: dit/dit.py [hunk 4]
```diff
@@ -327,90 +359,73 @@ def _export_all(self, concluded, verbose):
...
+        task = None
 
-        if task_id:
-            task = self.index[group_id][1][subgroup_id][1][task_id]
-        else:
-            task = None
+        if subgroup_id:
+            subgroup = self.index[group_id][1][subgroup_id][0]
+            if task_id:
+                task = self.index[group_id][1][subgroup_id][1][task_id]
```
Comment (danielmoraes) [inline dit/dit.py:410]: Yeah. Current group/subgroup does make sense on gid/gname parsing.

File: dit/dit.py [hunk 5]
```diff
@@ -424,47 +439,52 @@ def _id_parse(self, argv):
...
         names = [None] * (3 - len(names)) + names
+
         group, subgroup, task = names
 
         if group is None:
-            group = self.current_group
-        elif group == "":
-            group = None
+            if self.current_group is not None:
+                group = self.current_group
```
Comment (danielmoraes) [inline dit/dit.py:465]: Why default values for current_group, current_subgroup is not ""? Makes more sense and this checking is not necessary.

**Comment by danielmoraes** [comment]
Maybe "." is not necessary, since `dit list ""` worked on my machine. Also, I think it is better to use "../" or ".." to list current group and "./" or "./" to list current subgroup.

File: dit/dit.py (hunk 1 above)
Comment (filipelbc) [reply dit/dit.py:288]: Someone might create a directory in the wrong place... It does not hurt to check.

Trivial checks like this should be done as much as possible.

File: dit/dit.py (hunk 5 above)
Comment (filipelbc) [reply dit/dit.py:465]: `None` should be meaningful and semantically different from `""`.

The bug was caused because we were using `None` instead of `""` in the first place.

The function **should be able to return `None` to indicate the absence of a value, and `""` to indicate the root**.

**Comment by filipelbc** [comment]
Yes, now it works. Strange. But it does not work in the tests...

File: dit/dit.py (hunk 3 above)
Comment (filipelbc) [reply dit/dit.py:413]: This is intended. The code will fail `if (sub)group_id == None and current_(sub)group == None`. I expect some wrong index exception. I will create an issue for properly handling errors.

**Comment by filipelbc** [comment]
Write a proper proposal (issue and description) for this "./" thing. Put label "thoughts".

File: dit/dit.py (hunk 5 above)
Comment (danielmoraes) [reply dit/dit.py:465]: `None` and `""` were already being treated differently in previous versions. That is not the point. The problem is that current_(sub)group is not consistent. What a current_(sub)group equals to None means? Currently, it means that you have never worked in any task, but makes no sense. I believe the initial value for current_(sub)group should be "" (the root group and subgroup). In that way, this method will always return some valid (sub)group and the checking is not necessary.

Why current_(sub)group is currently **inconsistent** and its initial state should not be `None`? In the current implementation, current_(sub)group is None if and only if you never worked in any task. That is, after your first `dit workon`, current_(sub)group will never be None again. Instead, it will be pointing to the last (sub)grouped you worked on. Now assume that you have created some tasks, worked and concluded all of them. There are now zero pending tasks, all of them are concluded. This state is similar to the initial one (not working in any task), but one is None and the other is not.

Comment (filipelbc) [reply dit/dit.py:465]: This is exactly the intended behavior. You can create an issue for discussing that.
